### PR TITLE
Add SkippedTargetEventArgs and BuildReason

### LINKED ISSUE #2751: Extend TargetStartedEventArgs with dependency info
- [ ] On TargetStartedEventArgs: enum Reason { BeforeTargets, DependsOn, AfterTargets } + use ParentTarget
 - [ ] On TargetSkippedEventArgs: same Reason

### PR DESCRIPTION
**Pull request by panopticoncentral**
This adds:

* SkippedTargetEventArgs which provides additional logging information about targets skipped during build.
* BuildReason for both SkippedTargetEventArgs and TargetStartedEventArgs that indicates whether the target was built because of its parent's BeforeTarget, DependsOn, or AfterTarget.

This fixes #2751. It also bumps the version of binlog.

cc: @AndyGerlicher @KirillOsenkov

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
Very nice! Glad it turned out to be pretty straightforward.

**Comment by KirillOsenkov** [comment]
Are you planning to also add the TargetFile which is the file in which the target is defined? For https://github.com/KirillOsenkov/MSBuildStructuredLog/issues/117

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 1]
```diff
@@ -269,6 +292,7 @@ private BuildEventArgs ReadTargetStartedEventArgs()
             var projectFile = ReadOptionalString();
             var targetFile = ReadOptionalString();
             var parentTarget = ReadOptionalString();
+            var buildReason = (TargetBuiltReason) ReadInt32();
```
Comment (KirillOsenkov) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: If we're reading an older .binlog, don't attempt to read this field.

This should be:

```
TargetBuiltReason buildReason = TargetBuiltReason.None;
if (fileFormatVersion > 3)
{
    buildReason = (TargetBuiltReason)ReadInt32();
}
```

**Comment by AndyGerlicher** [comment]
@dotnet-bot test this please

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please
@dotnet-bot test Windows_NT Build for Full please

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please
@dotnet-bot test Windows_NT Build for Full please

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please
@dotnet-bot test Windows_NT Build for Full please

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please
@dotnet-bot test Windows_NT Build for Full please

**Comment by panopticoncentral** [comment]
Also added the TargetFile in.

**Comment by AndyGerlicher** [comment]
FYI when you change the public interface you have to build for netcore (`.\build.cmd /t:Build /p:Configuration=Debug-NetCore`) so it will update the refs. I pushed a commit that updates that though.

File: src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs [hunk 2]
```diff
@@ -358,6 +360,30 @@ public void RoundtripProjectImportedEventArgs()
...
+        public void RoundtripTargetSkippedEventArgs()
+        {
+            var args = new TargetSkippedEventArgs(
+                "Message")
+            {
+                BuildEventContext = BuildEventContext.Invalid,
+                ProjectFile = "foo.csproj",
+                ParentTarget = "bar",
+                BuildReason = TargetBuiltReason.DependsOn
+            };
```
Comment (KirillOsenkov) [inline src/Build.UnitTests/BuildEventArgsSerialization_Tests.cs]: I've just realized that @panopticoncentral @AndyGerlicher we forgot to serialize the TargetName. I think we'll need a separate PR that bumps the format version again.

Or, @AndyGerlicher you could sneak it in file format version 5 together with your profile changes?

Just need to read and write the TargetName for TargetSkippedEventArgs.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 3]
```diff
@@ -162,6 +165,27 @@ private BuildEventArgs ReadProjectImportedEventArgs()
...
         }
 
+        private BuildEventArgs ReadTargetSkippedEventArgs()
+        {
+            var fields = ReadBuildEventArgsFields();
+            // Read unused Importance, it defaults to Low
+            ReadInt32();
+            var targetFile = ReadOptionalString();
+            var parentTarget = ReadOptionalString();
+            var buildReason = (TargetBuiltReason)ReadInt32();
```
Comment (KirillOsenkov) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: This is where we forgot to read the TargetName.

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 4]
```diff
@@ -266,6 +273,15 @@ private void Write(ProjectImportedEventArgs e)
             WriteOptionalString(e.UnexpandedProject);
         }
 
+        private void Write(TargetSkippedEventArgs e)
+        {
+            Write(BinaryLogRecordKind.TargetSkipped);
+            WriteMessageFields(e);
+            WriteOptionalString(e.TargetFile);
+            WriteOptionalString(e.ParentTarget);
+            Write((int)e.BuildReason);
```
Comment (KirillOsenkov) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs]: We should write the TargetName here.

**Pull request merged**